zilliqa/src/node.rs

```diff
@@ -263,26 +263,66 @@ impl Node {
         self.consensus.head_block().header.number
     }
 
-    pub fn resolve_block_number(&self, number: BlockNumberOrTag) -> u64 {
-        // TODO(#863): This doesn't respect two-chain finalization.
-        match number {
-            BlockNumberOrTag::Latest => self.get_chain_tip(),
-            BlockNumberOrTag::Finalized => self.get_chain_tip().saturating_sub(2),
-            BlockNumberOrTag::Safe => self.get_chain_tip().saturating_sub(2),
+    pub fn resolve_block_number(&self, block_number: BlockNumberOrTag) -> u64 {
+        match block_number {
+            BlockNumberOrTag::Number(n) => n,
             BlockNumberOrTag::Earliest => 0,
+            BlockNumberOrTag::Latest => self.get_chain_tip(),
             BlockNumberOrTag::Pending => self.get_chain_tip(),
-            BlockNumberOrTag::Number(n) => n,
+            BlockNumberOrTag::Finalized => {
+                let Ok(Some(view)) = self.db.get_latest_finalized_view() else {
+                    return 0u64;
```

Comment:
Let's not use 0u64 as an error indicator? Sooner or later it will escape from this function and we will get code attempting to reference (or return) block 0x0, not 0x .

Reply:
I think you can reasonably and correctly argue that the genesis block (with view 0) is always finalized. In fact I would probably make that argument by changing the signature of `get_latest_finalized_view` to `Result<u64>` and returning 0 if there is nothing in the DB, along with a comment explaining why.

Reply:
Oh there are two levels of pattern matching here - I would argue handling mapping `None` to `0` is valid, but `Err` to `0` isn't. Make `resolve_block_number` return `Result<u64>` if its fallible.

Reply:
I may just be being overly pedanting, but I think that `None` makes sense if there are no blocks in the database. Genesis always being finalized sounds correct, but if you don't even know/have genesis, then your "latest finalized block" isn't genesis, it's none.

Reply:
Maybe we should document this invariant more clearly, but I'm 99% sure its impossible for `Consensus` to exist with fewer than one block (i.e. genesis always exists), therefore there is always a finalised block.

Reply:
That's a good point. Then we can remove the optionality from a few database methods. (Probably in another PR, no need to hold up this one)